File: src/main/java/ling572/ModelExpectation.java
package ling572;

import java.util.*;

import ling572.util.Instance;

public class ModelExpectation extends Expectation {
	private MaxEntModel model = null;
	private double defaultClassProb = 0; 
	
	@Override
	protected double getCount(String y, String x) {
		double exp = this.getExpectation(y, x);
		double count = exp * this.instanceCount;
		return count;
	}

	@Override
	public void build() {
		List<String> classes = this.getClasses();
		
		this.defaultClassProb = 1d/this.getClasses().size();
		
		
		for (Instance x : this.instances) {
			Map<String, Double> classProbs = null;
			
			if (this.model != null)
				classProbs = this.model.scoreInstanceJava(x);
			
			for (String t : x.getFeatures().keySet()) {
				for (String y : classes){					
					double exp;
					
					if (this.model == null)
						exp = this.calcExpectation(y, t);
					else {
						double classProb = classProbs.get(y);
						exp = this.calcExpectation(y, t, classProb);
					}

					this.addExp(y, t, exp);
				}
			}
		}

	}
	
	public void setMaxEntModel(MaxEntModel model) {
		this.model = model;
	}

	@Override
	protected double calcExpectation(String y, String t) {
		return this.calcExpectation(y, t, this.defaultClassProb);
	}

	protected double calcExpectation(String y, String t, double classProb) {
		return 1d/this.instanceCount * classProb; 
	}
	
	private void addExp(String y, String x, double exp) {
		double totalExp = this.getExpectation(y, x);
		totalExp += exp;		
		this.setExpectation(y, x, totalExp);
	}
}

File: src/main/scala/ling572/port/ModelExpectation.scala
package ling572.port

import ling572.MaxEntModel
import collection.immutable.ListMap
import collection.JavaConverters._


class ModelExpectation extends Expectation {

  private var model:MaxEntModel = null
  private var defaultClassProb = 0.0

  override protected def getCount(y:String, x:String) = {
    val exp = this.getExpectation(y, x)
    exp * this.instanceCount
  }

  override def build() {
    val classes = this.getClasses

    this.defaultClassProb = 1d/this.getClasses.size

    for (x <- this.instances) {
      val classProbs: ListMap[String, Double] = if (model != null ) model.scoreInstance(x)._2  else null

      for (t <- x.getFeatures.asScala.keys) {
        for (y <- classes){
          val exp = if (this.model == null)
            this.calcExpectation(y, t)
          else {
            this.calcExpectation(y, t, classProbs.getOrElse(y,0.0))
          }

          this.addExp(y, t, exp)
        }
      }
    }

  }

  def setMaxEntModel(model:MaxEntModel) {
    this.model = model
  }


  override protected def calcExpectation(y:String, t:String ) = {
    this.calcExpectation(y, t, this.defaultClassProb)
  }

  protected def calcExpectation(y:String, t:String, classProb:Double) = {
    1d/this.instanceCount * classProb
  }

  private def addExp(y:String, x:String, exp:Double) {
    val totalExp = this.getExpectation(y, x)
    this.setExpectation(y, x, totalExp + exp)
  }

}
